### ISSUE TITLE
Feat : Ajout des champs additionnels - Enhancement

### ISSUE THREAD

**Issue by Adrien-Pajot** [PnX-SI/gn_mobile_occtax#122]
Demande d'un BE (Emberiza) 
Je ne sais pas si cela a déjà été discuté mais pour suivre les évolutions d'OccTax, il pourrait être intéressant d'avoir une correspondance au niveau des champs additionnels. Chaque champ ajouté en web se retrouverait en mobile, ou au moins il serait possible d'en paramétrer sur les mobiles et puis de faire une correspondance ?

**Comment by DonovanMaillard** [comment]
C'est prévu et ca sera fait courant 2022. Nos prochaines commandes (j'ai.un cahier des charges à finir pour lundi prochain ;) ) vont commencer à réaligner le module Web et le module mobile dont on veut qu'ils soient les plus cohérents possibles, et que le mobile ne soit qu'un support différent mais avec des fonctionnalités a l'identique. 

Dans ce cadre la, la saisie en texte des observateurs, les champs additionnels, le lien entre dataset/liste de taxons etc sont prévus. Possiblement l'arrivée des médias (pas inclu dans notre previsionnel, à voir selon reliquats), voire des donnees en ligne/polygones.

**Comment by camillemonchicourt** [comment]
En effet, depuis plusieurs versions de GeoNature, des champs additionnels peuvent être ajoutés dans Occtax, globalement ou par JDD.
Il est possible d'ajouter des champs additionnels au niveau du relevé, au niveau du taxon ou au niveau du dénombrement.
Détail sur https://docs.geonature.fr/admin-manual.html#administration-des-champs-additionnels.
Les valeurs renseignées au niveau de chaque observation pour ces éventuelles champs additionnels sont stockées dans des champs JSON au niveau des tables des relevés, des taxons ou des dénombrements.

Les champs additionnels sont récupérables dans l'API : https://demo.geonature.fr/geonature/api/gn_commons/additional_fields

**Comment by TheoLechemia** [comment]
Bonjour Sébastien,
voici quelques réponses : 

> Le champ additionnel peut préciser le label (alors qu'on a déjà le label coté nomenclature)_


Effectivement. La on va privilégier le label fourni par le champs additionnel (une nomenclature pouvant être utilisée dans différents contextes, on veut parfois préciser son utilisation dans le contexte du protocole / formulaire)

> L'attribut field_values peut être renseigné ce qui peut être en conflit avec les valeurs possibles issues de la nomenclature.

Effectivement, le backoffice est très générique et on peut renseigner le "valeurs" même si on a dit que c'était un champs de type "nomenclature". Ces infos ne sont pas à prendre en compte : si le type de widget est "nomenclature" alors on ne prend que les valeurs renvoyé par l'API des nomenclatures

_L'attribut field_name sert à faire le mapping et est utilisé comme clé de la propriété à stocker coté relevé. Il diffère du code utilisé coté nomenclature._
Oui c'est bien de le champs `field_name` qu'il faut utiliser comme clé de champs JSON `additional_fields` de la table correspondante. Je ne comprend pas le "il diffère du code utilisé côté nomenclature" ?
Pour ce champs, la liste déroulante doit être constituée des valeurs que contient la liste associé au code nomenclature du champs additionnel. Ce qui nécessite que cette nomenclature ai été synchronisée au préalable côté mobile j'imagine ?

> Coté mapping, idéalement, ce serait de toujours s'appuyer sur l'attribut field_name comme clé de la propriété à ajouter au relevé.

oui

De ce que je lis, il y a aussi un élément qui semble manquant : 
il y a des champs additionnels globaux (à afficher tout le temps), et des champs additionnels associé à des jeux de données (à n'afficher que se le JDD est sélectionné).  Côté API, si le tableau `datasets` est vide, c'est que c'est un champs global, sinon il n'est à afficher que sur les JDD présent dans ce tableau

**Comment by TheoLechemia** [comment]
Autre vigilance : 
depuis la version 2.12, on a améliorer le "typage" du champs `field_values`. Pour les widget de type select, multiselect, checkbox et radio, le champs `field_values` doit contenir un tableau de dictionnaire avec les clés `label` et `value`. Il n'est plus possible de renseigner ce champs avec un simple tableau de valeur, mais on a en revanche pas reprise les champs existants. Du coup côté frontend, il y a une transformation des tableau en tableau de dictionnaire avec les clés `label` et `value`. J'imagine qu'il faudra que tu fasse pareil côté mobile

**Comment by sgrimault** [comment]
Bonjour @TheoLechemia,

Merci pour tes retours.
Je suis en train de revoir le modèle pour associer les champs additionnels avec les jeux de données. On pourra alors appliquer des filtres selon le jeu de données.
Pour la gestion des valeurs possibles (via l'attribut `field_values`), je construis systématiquement une liste de couples "clé", "valeur" où la valeur prend automatiquement le nom de la clé si aucune valeur n'a été définie. Le modèle accepte actuellement ces deux constructions (liste de valeurs simples, ou liste de couples).

Concernant la partie nomenclature, je me disais si ce ne serait pas plus simple de déclarer en plus dans la configuration les nomenclatures que l'on souhaite afficher pour la partie "Information" et "Dénombrement". Actuellement, on peut configurer cette liste via le paramètre `nomenclature` (cf. [Nomenclature settings](https://github.com/PnX-SI/gn_mobile_occtax#nomenclature-settings)). Actuellement cette liste est fermée et n'accepte pas des déclarations supplémentaires. Mais on pourrait imaginer de la rendre ouverte et ainsi ajouter les déclarations de la nomenclature que l'on souhaite afficher ou non (en les omettant).
Le souci de cette approche est qu'actuellement on a un mapping en dur sur les attributs à insérer dans le JSON du relevé. Exemple, le type de nomenclature `ETA_BIO` est traduit par `id_nomenclature_bio_condition` coté relevé… Idéalement, il faudrait tout simplement garder le code mnémonique de la nomenclature afin d'avoir un modèle souple et extensible.

L'autre approche serait de considérer que tous les champs des écrans "Information" et "Dénombrement" soient considérés comme des "champs additionnels" (avec le code du champ via l'attribut `field_name`, son libellé via l'attribut `field_label`, son type via l'attribut `widget_name`, etc.). Les types de nomenclature "éditables" seraient considérés comme un champ comme un autre, typé, avec ses propriétés. On aurait du coup quelque chose de totalement dynamique et pilotable par configuration et la construction de ces écrans serait uniforme.
Mais ça implique de revoir je pense des choses coté GeoNature, notamment au niveau du mapping.

**Comment by camillemonchicourt** [comment]
Dans la 2.7.0-rc1, un paramètre `nomenclature/additional_fields` a été ajouté (à `false` par défaut).
OK si c'est pour la phase de test, mais sinon, selon moi ce n'est pas utile de garder ce paramètre pour la release 2.7.0, car les champs additionnels s'affichent si certains ont été définis, ou pas si il n'y en a aucun définis au niveau d'Occtax ou d'un JDD spécifique.

**Comment by DonovanMaillard** [comment]
Oui, pas de raisons de les désactiver selon le terminal utilisé :)

**Comment by camillemonchicourt** [comment]
La 2.8 ajoute la prise en charge des champs additionnels au niveau des relevés.
Ainsi que leur  ordre, caractère obligatoire et leur valeur par défaut
Et le paramètre `additional_fields` a ainsi été passé à TRUE par défaut.

**Comment by camillemonchicourt** [comment]
@sgrimault, peux-tu faire un point ici sur les types de champs additionnels qui sont pris en charge parmi : 

- Select
- Checkbox
- Nomenclature
- Texte + Textarea
- Radio
- Time
- Date
- Multiselect
- Number
- Html

Et les propriétés qui sont prises en charge (obligatoire, valeur par défaut, etc... ?)

**Comment by sgrimault** [comment]
@camillemonchicourt,

Actuellement l'implémentation des champs additionnels supporte tous les champs listés sauf `Html`. Ce dernier implique un champ de saisie de type `Textarea` avec possibilité de mise en forme plus avancée (structure comme les titres, gestion des listes, tableaux, ajout d'images, autres ?). De mémoire, ça avait été écarté, mais ce qui pourrait être fait est de quand même afficher ce champ mais présenté comme un champ de type `Textarea` sans possibilité de mise en forme.
Parmi les options de saisie, on a le support du caractère obligatoire ou non du champ et de l'affichage d'une valeur par défaut (si défini).

**Comment by camillemonchicourt** [comment]
OK.
En effet les champs addition et de type HTML ne sont pas des champs à saisir ou renseigner. Ce ne sont pas des champs de texte riche.
Ce sont des blocs HTML d'info qui permettenr d'ajouter des infos dans le formulaire.

Donc un truc un peu spécial, en effet qui n'est pas à supporter dans Occtax-mobile.

Merci pour les précisions.
